File: 3116-kth-smallest-amount-with-single-denomination-combination/3116-kth-smallest-amount-with-single-denomination-combination.py
import math

class Solution:
    def findKthSmallest(self, coins: List[int], k: int) -> int:
        n = len(coins)
        lcm_list = []
        
        for i in range(1, 1 << n):
            current_lcm = 1
            count = 0

            for j in range(n):
                if (i >> j) & 1:
                    current_lcm = math.lcm(current_lcm, coins[j])
                    count += 1

            lcm_list.append((current_lcm, count))

        def count_amounts(x: int) -> int:
            res = 0
            for lcm_val, count in lcm_list:
                if count % 2 == 1:
                    res += x // lcm_val
                else:
                    res -= x // lcm_val
            return res

        low = 1
        high = min(coins) * k
        ans = high

        while low <= high:
            mid = (low + high) // 2

            if count_amounts(mid) >= k:
                ans = mid
                high = mid - 1
            else:
                low = mid + 1
        
        return ans

File: 3116-kth-smallest-amount-with-single-denomination-combination/3116-kth-smallest-amount-with-single-denomination-combination.cpp
class Solution {
public:
    long long findKthSmallest(vector<int>& coins, int k) {
        int n = coins.size();
        
        vector<pair<long long, int>> lcm_list;
        for (int i = 1; i < (1 << n); ++i) {
            long long current_lcm = 1;
            int count = 0;
            for (int j = 0; j < n; ++j) {
                if ((i >> j) & 1) {
                    current_lcm = std::lcm(current_lcm, (long long)coins[j]);
                    count++;
                }
            }
            lcm_list.push_back({current_lcm, count});
        }
        
        auto count_amounts = [&](long long x) {
            long long res = 0;
            for (const auto& p : lcm_list) {
                if (p.second % 2 == 1) { 
                    res += x / p.first;
                } else {                 
                    res -= x / p.first;
                }
            }
            return res;
        };
        
        long long low = 1;
        long long high = (long long)*min_element(coins.begin(), coins.end()) * k;
        long long ans = high;
        
        while (low <= high) {
            long long mid = low + (high - low) / 2;
            
            if (count_amounts(mid) >= k) {
                ans = mid;       
                high = mid - 1;
            } else {
                low = mid + 1;   
            }
        }
        
        return ans;
    }
};